File: typescript/src/model/ral_object.ts
import { ObjectRef } from "model/object_ref";
import { CurrentGeoLocation } from "./current_geo_location";
import { Definition } from "./definition";
import { Identity } from "./identity";
import { SpecificProperties } from "./specific_properties";
import { SpecificProperty } from "./specific_property";
import { Template } from "./template";


export { CurrentGeoLocation, Definition, Identity, SpecificProperties, Template, SpecificProperty };

export class RalObject {
    public identity: Identity;
    public currentOwners: ObjectRef[];
    public definition: Definition;
    public objectState: string;
    public template: Template;
    public specificProperties: SpecificProperties;
    public currentGeoLocation: CurrentGeoLocation;
    public locationHistoryRef: string[];
    public ownerHistoryRef: string[];
    public methodHistoryRef: string[];
    public linkedObjectRef: ObjectRef[];

    constructor(
        params: {
            identity: Identity,
            definition: Definition,
            template: Template,
            specificProperties: SpecificProperties,
            currentGeoLocation: CurrentGeoLocation,
            currentOwners?: ObjectRef[],
            objectState?: string,
            locationHistoryRef?: string[],
            ownerHistoryRef?: string[],
            methodHistoryRef?: string[],
            linkedObjectRef?: ObjectRef[],
        }
    ) {
        this.identity = params.identity;
        this.currentOwners = params.currentOwners || [];
        this.definition = params.definition;
        this.objectState = params.objectState || "undefined";
        this.template = params.template;
        this.specificProperties = params.specificProperties;
        this.currentGeoLocation = params.currentGeoLocation;
        this.locationHistoryRef = params.locationHistoryRef || [];
        this.ownerHistoryRef = params.ownerHistoryRef || [];
        this.methodHistoryRef = params.methodHistoryRef || [];
        this.linkedObjectRef = params.linkedObjectRef || [];
    }

    transformTo(specificPropertiesTransform: (specificProperties: SpecificProperties) => SpecificProperties): RalObject {
        return new RalObject({
            identity: this.identity,
            definition: this.definition,
            template: this.template,
            specificProperties: specificPropertiesTransform(this.specificProperties),
            currentGeoLocation: this.currentGeoLocation,
            currentOwners: this.currentOwners,
            objectState: this.objectState,
            locationHistoryRef: this.locationHistoryRef,
            ownerHistoryRef: this.ownerHistoryRef,
            methodHistoryRef: this.methodHistoryRef,
            linkedObjectRef: this.linkedObjectRef
        });
    }

    toMap(): Record<string, any> {
        return {
            identity: this.identity.toMap(),
            currentOwners: this.currentOwners.map(val => val.toMap()),
            definition: this.definition.toMap(),
            objectState: this.objectState,
            template: this.template.toMap(),
            specificProperties: this.specificProperties.toMaps(),
            currentGeoLocation: this.currentGeoLocation.toMap(),
            locationHistoryRef: this.locationHistoryRef,
            ownerHistoryRef: this.ownerHistoryRef,
            methodHistoryRef: this.methodHistoryRef,
            linkedObjectRef: this.linkedObjectRef.map(val => val.toMap()),
        };
    }

    static fromMap(map: Record<string, any>): RalObject {
        const identity: Identity = Identity.fromMap(map.identity);
        const currentOwners: ObjectRef[] = map.currentOwners != null ? map.currentOwners.map((val: any) => { return ObjectRef.fromMap(val) }) : [];
        const definition: Definition = Definition.fromMap(map.definition);
        const objectState: string = map.objectState || "undefined";
        const template: Template = Template.fromMap(map.template);
        const specificProperties: SpecificProperties = SpecificProperties.fromMaps(map.specificProperties);
        const currentGeoLocation: CurrentGeoLocation = CurrentGeoLocation.fromMap(map.currentGeoLocation);
        const locationHistoryRef: string[] = map.locationHistoryRef || [];
        const ownerHistoryRef: string[] = map.ownerHistoryRef || [];
        const methodHistoryRef: string[] = map.methodHistoryRef || [];
        const linkedObjectRef: ObjectRef[] = map.linkedObjectRef != null ? map.linkedObjectRef.map((val: any) => { return ObjectRef.fromMap(val) }) : [];

        return new RalObject({
            identity: identity,
            definition: definition,
            template: template,
            specificProperties: specificProperties,
            currentGeoLocation: currentGeoLocation,
            currentOwners: currentOwners,
            objectState: objectState,
            locationHistoryRef: locationHistoryRef,
            ownerHistoryRef: ownerHistoryRef,
            methodHistoryRef: methodHistoryRef,
            linkedObjectRef: linkedObjectRef,
        });
    }

    //simple to string method
    toString(): string {
        return `RalObject(${this.identity.uid}, ${JSON.stringify(this)})`
    }
}

File: dart/lib/src/model/ral_object.dart
import 'package:either_dart/either.dart';
import 'package:openral_core/src/cross/backend/parsing/parsing_error.dart';
import 'package:openral_core/src/model/current_geo_location.dart';
import 'package:openral_core/src/model/definition.dart';
import 'package:openral_core/src/model/identity.dart';
import 'package:openral_core/src/model/object_ref.dart';
import 'package:openral_core/src/model/object_template.dart';
import 'package:openral_core/src/model/specific_properties.dart';
import 'package:openral_core/src/model_parser/ral_object_parser.dart';

///represents the base class for RAL objects
///Represents a json object like this:
///```json
/// {
///   "identity": {
///     "UID": "",
///     "name": "thing",
///     "siteTag": "",
///     "alternateIDs": [],
///     "alternateNames": []
///   },
///   "currentOwners": [],
///   "definition": {
///     "definitionText": "An object or entity that is not or cannot be named specifically",
///     "definitionURL": "https://www.thefreedictionary.com/thing"
///   },
///   "objectState": "undefined",
///   "template": {
///     "RALType": "thing",
///     "version": "1",
///     "objectStateTemplates": ["generalObjectState"]
///   },
///   "specificProperties": [
///     {
///       "key": "serial number",
///       "value": "",
///       "unit": "String"
///     }
///   ],
///   "currentGeolocation": {
///     "container": {
///       "UID": "unknown"
///     },
///     "postalAddress": {
///       "country": "unknown",
///       "cityName": "unknown",
///       "cityNumber": "unknown",
///       "streetName": "unknown",
///       "streetNumber": "unknown"
///     },
///     "3WordCode": "unknown",
///     "geoCoordinates": {
///       "longitude": 0,
///       "latitude": 0
///     },
///     "plusCode": "unknown"
///   },
///   "locationHistoryRef": [],
///   "ownerHistoryRef": [],
///   "methodHistoryRef": [],
///   "linkedObjectRef": []
/// }
///
///```
class RalObject<S extends SpecificProperties> {
  //Note: The fields are not final, because otherwise we would have to create a copyWith constructor for every nested field,
  //which would be a lot of boilerplate code

  Identity identity;

  List<ObjectRef> currentOwners;

  Definition definition;

  String objectState;

  ObjectTemplate template;

  S _specificProperties;

  CurrentGeoLocation currentGeoLocation;

  List<String> locationHistoryRef;

  List<String> ownerHistoryRef;

  List<String> methodHistoryRef;

  List<ObjectRef> linkedObjectRef;

  RalObject({
    required this.identity,
    this.currentOwners = const [],
    required this.definition,
    this.objectState = "undefined",
    required this.template,
    required S specificProperties,
    required this.currentGeoLocation,
    this.locationHistoryRef = const [],
    this.ownerHistoryRef = const [],
    this.methodHistoryRef = const [],
    this.linkedObjectRef = const <ObjectRef>[],
  }) : _specificProperties = specificProperties;

  ///by defining a getter, we can override SpecificProperties in subclasses
  S get specificProperties => _specificProperties;

  static Either<RalObject<S>, ParsingError> fromMap<S extends SpecificProperties>(
    Map<String, dynamic> map, {
    S Function(SpecificProperties specificProperties)? specificPropertiesTransform,
  }) {
    final parserFactory = RalObjectParser();

    final parsingResult = parserFactory.fromMap(map);

    if (parsingResult.isRight) {
      //returns the parsing error
      return Right(parsingResult.right);
    } else {
      final originalObject = parsingResult.left;

      if (specificPropertiesTransform != null) {
        //returns the transformed object
        final transformed = originalObject.transformTo<S>(specificPropertiesTransform);

        return Left(transformed);
      } else {
        return Left(parsingResult.left as RalObject<S>);
      }
    }
  }

  ///returns this [RalObject] to a RalObject with the given [SpecificProperties] subclass by using the given [specificPropertiesTransform] function
  RalObject<S> transformTo<S extends SpecificProperties>(
    S Function(SpecificProperties specificProperties) specificPropertiesTransform,
  ) {
    return RalObject<S>(
      identity: identity,
      currentOwners: currentOwners,
      definition: definition,
      objectState: objectState,
      template: template,
      specificProperties: specificPropertiesTransform(specificProperties), //transform the specific properties
      currentGeoLocation: currentGeoLocation,
      locationHistoryRef: locationHistoryRef,
      ownerHistoryRef: ownerHistoryRef,
      methodHistoryRef: methodHistoryRef,
      linkedObjectRef: linkedObjectRef,
    );
  }

  Map<String, dynamic> toMap() {
    return {
      "identity": identity.toMap(),
      "currentOwners": currentOwners.map<Map<String, dynamic>>((e) => e.toMap()).toList(),
      "definition": definition.toMap(),
      "objectState": objectState,
      "template": template.toMap(),
      "specificProperties": specificProperties.toMaps(),
      "currentGeolocation": currentGeoLocation.toMap(),
      "locationHistoryRef": locationHistoryRef,
      "ownerHistoryRef": ownerHistoryRef,
      "methodHistoryRef": methodHistoryRef,
      "linkedObjectRef": linkedObjectRef.map<Map<String, dynamic>>((e) => e.toMap()).toList(),
    };
  }
}
